packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -1641,20 +1641,30 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     _lastBottomViewInset = WidgetsBinding.instance.window.viewInsets.bottom;
   }
 
+  _WhitespaceDirectionalityFormatter _whitespaceFormatter;
+
   void _formatAndSetValue(TextEditingValue value) {
+    _whitespaceFormatter ??= _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
+
     // Check if the new value is the same as the current local value, or is the same
     // as the post-formatting value of the previous pass.
     final bool textChanged = _value?.text != value?.text;
-    final bool isRepeat = value?.text == _lastFormattedUnmodifiedTextEditingValue?.text;
-    if (textChanged && !isRepeat && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
-      for (final TextInputFormatter formatter in widget.inputFormatters)
-        value = formatter.formatEditUpdate(_value, value);
-      _value = value;
-      _updateRemoteEditingValueIfNeeded();
-    } else {
+    final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
+    // Always pass the text through the whitespace directionality formatter to maintain
+    // expected behavior with carets on trailing whitespace.
+    if (!isRepeat) {
+      if (textChanged && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
+        for (final TextInputFormatter formatter in widget.inputFormatters) {
+          value = formatter.formatEditUpdate(_value, value);
+        }
+        value = _whitespaceFormatter.formatEditUpdate(_value, value);
+      }
       _value = value;
     }
-    if (textChanged && widget.onChanged != null)
+    _updateRemoteEditingValueIfNeeded();
```

Comment:
From the platform's perspective, 2 "current editing values" have been set right? Once with what the user typed and once with the re-formatted values. Could there be other side effects? For instance, could the keyboard's word completion suggestions flicker between 2 values depending on whether you're showing suggestions for the word on the left side or the right side?

Reply:
I don't think a properly implemented system should experience this. This extra call to _updateRemoteEditingValueIfNeeded is an effort to sync the remote and local values to the same thing. If it changes, one of the versions will end up being cached in _value, which is treated as ground truth. Obviously, if the formatter was not deterministic or had some cached counter or something, then you could achieve flickering/runaway behavior, but that is just an improper impl.